In [99]:
!git clone https://github.com/emilymacris/wrangling.git


fatal: destination path 'wrangling' already exists and is not an empty directory.


# Assignment: Data Wrangling
## `! git clone https://github.com/DS3001/wrangling`
## Do Q2, and one of Q1 or Q3.

**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [100]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [101]:
df_1 = pd.read_csv('./wrangling/assignment/data/airbnb_hw.csv')
df_1.head(10)



Host Id Host Since                                 Name Neighbourhood   \
0   5162530        NaN      1 Bedroom in Prime Williamsburg       Brooklyn   
1  33134899        NaN      Sunny, Private room in Bushwick       Brooklyn   
2  39608626        NaN                 Sunny Room in Harlem      Manhattan   
3       500  6/26/2008   Gorgeous 1 BR with Private Balcony      Manhattan   
4       500  6/26/2008             Trendy Times Square Loft      Manhattan   
5      1039  7/25/2008   Big Greenpoint 1BD w/ Skyline View       Brooklyn   
6      1783  8/12/2008                         Amazing Also      Manhattan   
7      2078  8/15/2008  Colorful, quiet, & near the subway!       Brooklyn   
8      2339  8/20/2008  East Village Cocoon: 2 Bedroom Flat      Manhattan   
9      2339  8/20/2008   Lovely 2 Bedroom East Village Nest      Manhattan   

  Property Type  Review Scores Rating (bin)        Room Type  Zipcode  Beds  \
0     Apartment                         NaN  Entire home/apt  11249.0   1.0   
1     Apartment                         NaN     Private room  11206.0   1.0   
2     Apartment                         NaN     Private room  10032.0   1.0   
3     Apartment                         NaN  Entire home/apt  10024.0   3.0   
4     Apartment                        95.0     Private room  10036.0   3.0   
5     Apartment                       100.0  Entire home/apt  11222.0   1.0   
6     Apartment                       100.0  Entire home/apt  10004.0   1.0   
7     Apartment                        90.0     Private room  11201.0   1.0   
8     Apartment                        90.0  Entire home/apt  10009.0   2.0   
9     Apartment                        95.0  Entire home/apt  10009.0   2.0   

   Number of Records  Number Of Reviews Price  Review Scores Rating  
0                  1                  0   145                   NaN  
1                  1                  1    37                   NaN  
2                  1                  1    28                   NaN  
3                  1                  0   199                   NaN  
4                  1                 39   549                  96.0  
5                  1                  4   149                 100.0  
6                  1                  9   250                 100.0  
7                  1                 80    90                  94.0  
8                  1                 95   270                  90.0  
9                  1                 23   290                  96.0

In [102]:

# Converted 'Price' to string type so that we can clean i
df_1['Price'] = df_1['Price'].astype(str)

# Removed any non-numeric characters and convert the type of the Pricecolumn to float
df_1['Price'] = df_1['Price'].str.replace('[^\d.]', '', regex=True)

In [103]:
# coerced the column to being numeric
df_1['Price'] = pd.to_numeric(df_1['Price'], errors='coerce')

# Create a missing value dummy variable
df_1['price_nan'] = df_1['Price'].isnull().astype(int)

# Checked for any missing values after we cleaned the Price column
total_missing = df_1['price_nan'].sum()

print(total_missing)

0


QUESTION #2,
1. ANSWER:
I cleaned the Price variable in the airbnb_hw.csv dataset by first converting it to a string to remove non-numeric characters like dollar signs and commas. After cleaning, I converted the values to numeric, ensuring that any invalid entries became NaN to maintain data quality. I also created a dummy variable to track missing values.

After the cleaning process, I found 0 missing values. The Price column is now of type float, and the values are correctly formatted for analysis, with prices like "1,112" appearing as 1112.0. This thorough cleaning ensures the data is ready for further exploration.

In [104]:
df_2 = pd.read_csv('./wrangling/assignment/data/sharks.csv')
df_2.head(10)



<ipython-input-104-899a07ae6492>:1: DtypeWarning: Columns (10,17,18,19,20,21,24) have mixed types. Specify dtype option on import or set low_memory=False.
  df_2 = pd.read_csv('./wrangling/assignment/data/sharks.csv')


index   Case Number                  Date    Year          Type  \
0      0    2020.02.05           05-Feb-2020  2020.0    Unprovoked   
1      1  2020.01.30.R  Reported 30-Jan-2020  2020.0      Provoked   
2      2    2020.01.17           17-Jan-2020  2020.0    Unprovoked   
3      3    2020.01.16           16-Jan-2020  2020.0    Unprovoked   
4      4    2020.01.13           13-Jan-2020  2020.0    Unprovoked   
5      5    2020.01.12           12-Jan-2020  2020.0    Unprovoked   
6      6    2020.01.07           07-Jan-2020  2020.0    Unprovoked   
7      7    2020.01.05           05-Jan-2020  2020.0    Unprovoked   
8      8    2020.01.03           03-Jan-2020  2020.0    Unprovoked   
9      9    2020.01.02           02-Jan-2020  2020.0  Questionable   

         Country                Area                 Location  \
0            USA                Maui                      NaN   
1        BAHAMAS              Exumas                      NaN   
2      AUSTRALIA     New South Wales            Windang Beach   
3    NEW ZEALAND           Southland              Oreti Beach   
4            USA      North Carolina    Rodanthe, Dare County   
5       THAILAND  Phang Nga Province         Nang Thong Beach   
6      AUSTRALIA          Queensland        North West Island   
7      AUSTRALIA   Western Australia  Cull Island / Esperance   
8  NEW CALEDONIA                 NaN                   Calana   
9      AUSTRALIA   Western Australia                Esperance   

                               Activity               Name Unnamed: 9  Age  \
0              Stand-Up Paddle boarding                NaN        NaN  NaN   
1                              Floating    Ana Bruna Avila          F   24   
2                               Surfing     Will Schroeter          M   59   
3                               Surfing        Jordan King          F   13   
4                               Surfing       Samuel Horne          M   26   
5                              Swimming  Hans-Peter Malten          M   75   
6                              Swimming        Miah Holmes          F    7   
7                          Scuba diving       Gary Johnson          M   57   
8                                   NaN          Peter ___          M   12   
9  Swimming after being swept off rocks    Eric Birighitti          M   21   

                                      Injury Fatal (Y/N)       Time  \
0          No injury, but paddleboard bitten           N      09h40   
1  PROVOKED INCIDENT Scratches to left wrist           N        NaN   
2          Laceration ot left ankle and foot           N      08h00   
3                  Minor injury to lower leg           N      20h30   
4                        Lacerations to foot           N      14h33   
5                                Calf bitten           N      07h15   
6                         Lacerations to leg           N      17h30   
7                                      FATAL           Y      13h00   
8                           Right leg bitten           N  Afternoon   
9           Probable drowning and scavenging           Y      16h10   

                      Species   \
0                  Tiger shark   
1                          NaN   
2              "A small shark"   
3  Broadnose seven gill shark?   
4                          NaN   
5                          NaN   
6                  Lemon shark   
7                  White shark   
8                          NaN   
9                          NaN   

                              Investigator or Source  \
0                    K. McMurray, TrackingSharks.com   
1                    K. McMurray, TrackingSharks.com   
2  B. Myatt & M. Michaelson, GSAF; K. McMurray, T...   
3                    K. McMurray, TrackingSharks.com   
4                                  C. Creswell, GSAF   
5                    K. McMurray, TrackingSharks.com   
6  B. Myatt & M. Michaelson, GSAF; K. McMurray, T...   
7                                     B. Myatt, GSAF

In [105]:
# View the unique values in the 'Type' column

unique_types = df_2['Type'].unique()
print(unique_types)

['Unprovoked' 'Provoked' 'Questionable' 'Watercraft' 'Unconfirmed'
 'Unverified' 'Invalid' 'Under investigation' 'Boating' 'Sea Disaster' nan
 'Boat' 'Boatomg']


In [106]:
# Clean the 'Type' variable
df_2['Type'] = df_2['Type'].str.strip()  # Remove leading/trailing whitespace
df_2['Type'] = df_2['Type'].str.lower()  # Convert to lowercase for consistency
df_2['Type'] = df_2['Type'].str.replace(' ', '_')  # Replace spaces with underscores

# Replace missing values with NaN if needed
df_2['Type'] = df_2['Type'].replace(['', 'nan', 'NaN'], np.nan)

In [107]:
# Further cleaning can be done based on specific requirements, such as grouping similar types together.
# For example:
df_2['Type'] = df_2['Type'].replace(['boat', 'boating', 'Boat', 'Boating', 'boatomg', 'watercraft'], 'boat_related')

# Check the unique values after cleaning
cleaned_unique_types = df_2['Type'].unique()
print(cleaned_unique_types)



['unprovoked' 'provoked' 'questionable' 'boat_related' 'unconfirmed'
 'unverified' 'invalid' 'under_investigation' 'sea_disaster' nan]


In [108]:
# changed all values for Type that arent unprovoked, provoked, or boat_related to 'unverified'

df_2['Type'] = np.where(df_2['Type'].isin(['unprovoked', 'provoked', 'boat_related']), df_2['Type'], 'unverified')


In [109]:
cleaned_unique_types = df_2['Type'].unique()
print(cleaned_unique_types)

['unprovoked' 'provoked' 'unverified' 'boat_related']


Question #2
2. I cleaned the 'Type' variable in the sharks.csv dataset by first removing leading and trailing whitespaces and converting all entries to lowercase for ease. Then I replaced spaces with underscores to handle multi-word entries.
I grouped together all boating related types under'boat_related'. I then categorized all remaining types as 'unverified' to simplify the categories to make  only four unique values for type: 'unprovoked', 'provoked', 'boat_related', and 'unverified'.

In [141]:
df_3 = pd.read_parquet ('./wrangling/assignment/data/justice_data.parquet')


In [142]:
release = df_3['WhetherDefendantWasReleasedPretrial']

print(release.unique())



[9 0 1]


In [143]:
#df_3['WhetherDefendantWasReleasedPretrial'] = df_3['WhetherDefendantWasReleasedPretrial'].replace(['', ' ', 'nan', 'NaN'], np.nan)



In [144]:
release = release.replace(9, np.nan)
print(release.value_counts(),'\n')
# tells you how much are null
sum(release.isnull())
df_3['WhetherDefendantWasReleasedPretrial'] = release
del release

WhetherDefendantWasReleasedPretrial
1.0    19154
0.0     3801
Name: count, dtype: int64 



In [145]:
length = df_3['ImposedSentenceAllChargeInContactEvent']
type = df_3['SentenceTypeAllChargesAtConvictionInContactEvent']

# print( length.unique()  , '\n') # Some values are ' ', denoting missing
length = pd.to_numeric(length,errors='coerce') # Coerce to numeric
length_NA = length.isnull() # Create a missing dummy
print( np.sum(length_NA),'\n') # 9k missing values of 23k, not so good

print( pd.crosstab(length_NA, type), '\n') # Category 4 is cases where the charges were dismissed

length = length.mask( type == 4, 0) # Replace length with 0 when type ==4
length = length.mask( type == 9, np.nan) # Replace length with np.nan when type == 9

length_NA = length.isnull() # Create a new missing dummy
print( pd.crosstab(length_NA, type), '\n')
print( np.sum(length_NA),'\n') # 274 missing, much better

df_3['ImposedSentenceAllChargeInContactEvent'] = length # Replace data with cleaned version
del length, type # Delete temporary length/type variables

9053 

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
ImposedSentenceAllChargeInContactEvent                                      
False                                             8720  4299  914     0    0
True                                                 0     0    0  8779  274 

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
ImposedSentenceAllChargeInContactEvent                                      
False                                             8720  4299  914  8779    0
True                                                 0     0    0     0  274 

274 



**Q3.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?

Q3, Question #1
The 2020 U.S. Census made important updates to how it collects data on race and ethnicity:

A. New spaces for people to write in their specific identities were added for White and Black or African American respondents, allowing for more accurate answers.

B.  The Census now reports data for nearly 1,500 racial and ethnic groups. For example, there are 104 detailed groups for White respondents (like Dutch or Lebanese) and 62 for Black or African American respondents (such as Congolese or Grenadian). This is a major increase from previous censuses, which grouped many identities together.

C. The Census also included new categories for regional groups. This means data is now available for populations like Middle Eastern and North African, Caribbean, and Sub-Saharan African, reflecting the country's growing diversity.

D. Improved Data Collection: The changes in how data was collected and processed led to over 350 million responses to the race and ethnicity questions, which is six times more than in the 2010 Census. This allowed for much more detailed information on various groups.

Q3, Question #2
Gathering data on gender, race, etc helps us understand understand the demographics of the population, providing a clear picture of social dynamics and trends. This information is crucial for developing policies that meet the needs of diverse communities, ensuring that resources and services are distributed fairly. Additionally, accurate data plays a key role in monitoring and enforcing civil rights laws, helping to identify discrimination and inequality.

In politics, this data informs public debate by highlighting important issues such as access to education and healthcare disparities. It empowers communities to advocate for their needs and promotes the recognition of diverse cultures, fostering inclusivity. High-quality data also enhances decision-making. When policymakers rely on accurate information, they can allocate resources more effectively and create targeted interventions to address specific community needs.

Moreover, quality data builds trust among the public, policymakers, and researchers, ensuring that the conclusions drawn from the data are credible. It also allows for meaningful comparisons over time, helping us track progress and understand systemic issues. Overall, collecting and maintaining high-quality data on race and ethnicity is vital for creating a fair and equitable society.

Q3, Question #3:
Positive: One of the standout features of the 2020 Census was the introduction of expanded write-in options for race and ethnicity, allowing respondents to specify their identities more accurately. This change significantly increased the granularity of data collected, providing insights into a wider array of racial and ethnic groups. The adaptive design of data publication, which varied based on population size, also ensured that more detailed counts were available at various geographic levels. Additionally, the increased emphasis on data processing improvements resulted in a substantial rise in the number of responses, enhancing the overall data quality.
Negative: Despite these successes, the Census did not do well in obtaining data from hard-to-reach populations, like non-English speakers, rural residents, and homeless individuals. Increased outreach efforts, including partnerships with community organizations, could have improved participation rates in these groups. Moreover, while the Census expanded its categories for data collections, there are still communities that might not feel adequately represented. For instance, more detailed options for various Asian and Pacific Islander subgroups could have been included.





Q3, Question #4:
The 2020 Census gathered data on sex and gender through a straightforward question that asked respondents to identify as either male or female.

Positive: One positive aspect of the Census was its emphasis on clarity and simplicity in the question format. By asking a direct question about sex, the Census aimed to ensure high response rates and ease of understanding. Additionally, the Census Bureau provided guidance and resources to help respondents understand the question and its importance, promoting awareness of the data's role in policy-making.

Negative: Despite these strengths, the Census fell short in addressing the diversity of gender identities. The binary classification of male and female excluded non-binary, genderqueer, and transgender individuals, which can lead to significant underrepresentation in the data. Many respondents may have felt that their identities were not recognized, potentially discouraging participation. Furthermore, there was limited outreach to LGBTQ+ communities, which could have enhanced response rates and improved data quality.

Q3, Question #5
When cleaning data related to protected characteristics like sex, gender, sexual identity, and race, several concerns arise, particularly around missing values. One major issue is misclassification, where individuals may be incorrectly categorized, leading to inaccurate results. Privacy is also a significant concern; some individuals may feel uncomfortable disclosing their identities, resulting in gaps in data that can skew overall findings. Additionally, biases in the data cleaning process can perpetuate stereotypes and systemic inequities.

Missing values pose their own challenges. They can lead to a loss of important information, particularly if certain groups are underreported. Imputation methods, or filling in missing data, can introduce bias if not handled carefully—assuming binary gender for missing responses can erase non-binary identities, for example. Good practices include being transparent about how missing data is handled and providing inclusive options for individuals to specify their identities. Conversely, over-standardization and ignoring context can lead to high rates of missing data and misinterpretation. By addressing these challenges thoughtfully, we can improve the accuracy and inclusivity of the data collected.






I would be concerned with a few things if an algorithm were invented to impute values for protected characteristics like race, gender, sex, or sexuality. First, it could erase the complexity of individual identities by assigning categories based on data patterns that don’t necessarily reflect personal experiences. There is also a risk of reinforcing societal biases if the algorithm is trained on flawed historical data, leading to discriminatory outcomes. Additionally, the accuracy of the imputations might be questionable, and the algorithm may ignore important context, resulting in insensitive classifications. Finally, the lack of transparency in how the algorithm operates could undermine trust, and accountability for biased outcomes would be difficult to determine. To address these issues, it is essential to involve various parties in development to prioritize transparency, and regularly test the algorithm for biases.



